scio-hdfs/src/main/java/com/google/cloud/dataflow/contrib/hadoop/simpleauth/SimpleAuthHadoopFileSink.java

```diff
@@ -18,31 +18,154 @@
-  private final HadoopUserUtils user = new HadoopUserUtils();
+  private final String username;
 
   public SimpleAuthHadoopFileSink(String path,
                                   Class<? extends FileOutputFormat<K, V>> formatClass,
+                                  Configuration conf,
                                   String username) {
-    super(path, formatClass);
-    user.setSimpleAuthUser(username);
+    super(path, formatClass, conf);
+    this.username = username;
   }
 
-  public SimpleAuthHadoopFileSink(String path,
+  @Override
+  public WriteOperation<KV<K, V>, ?> createWriteOperation(PipelineOptions options) {
+    return new SimpleAuthHadoopWriteOperation<>(this, path, formatClass, username);
+  }
+
+  public static class SimpleAuthHadoopWriteOperation<K, V> extends HadoopWriteOperation<K, V> {
+    private final String username;
+
+    SimpleAuthHadoopWriteOperation(Sink<KV<K, V>> sink,
+                                          String path,
+                                          Class<? extends FileOutputFormat<K, V>> formatClass,
+                                          String username) {
+      super(sink, path, formatClass);
+      this.username = username;
+    }
+
+    @Override
+    public void finalize(Iterable<String> writerResults, PipelineOptions options) throws Exception {
+      final Iterable<String> results = writerResults;
+      final PipelineOptions opts = options;
+
+      UserGroupInformation.createRemoteUser(username).doAs(new PrivilegedExceptionAction<Void>() {
+        @Override
+        public Void run() throws Exception {
+          _finalize(results, opts);
+          return null;
+        }
+      });
+    }
+
+    private void _finalize(Iterable<String> writerResults, PipelineOptions options) throws Exception {
+      super.finalize(writerResults, options);
+    }
+
+    @Override
+    public Writer<KV<K, V>, String> createWriter(PipelineOptions options) throws Exception {
+      return new SimpleAuthHadoopWriter<>(this, path, formatClass, username);
+    }
+  }
+
+  public static class SimpleAuthHadoopWriter<K, V> extends HadoopWriter<K, V> {
+    private final UserGroupInformation ugi;
+
+    public SimpleAuthHadoopWriter(SimpleAuthHadoopWriteOperation<K, V> writeOperation,
+                                  String path,
                                   Class<? extends FileOutputFormat<K, V>> formatClass,
-                                  Configuration conf,
                                   String username) {
-    super(path, formatClass, conf);
-    user.setSimpleAuthUser(username);
+      super(writeOperation, path, formatClass);
+      ugi = UserGroupInformation.createRemoteUser(username);
+    }
+
+    @Override
+    public void open(String uId) throws Exception {
+      final String uid = uId;
+      ugi.doAs(new PrivilegedExceptionAction<Void>() {
+        @Override
+        public Void run() throws Exception {
+          _open(uid);
```

Comment:
I don't think that would work either because of dynamic method dispatch, right? Need to explicitly have the method in scope where child calls parent method